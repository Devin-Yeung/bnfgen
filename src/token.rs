use std::fmt::Display;
use std::{fmt::Formatter, num::ParseIntError};

use logos::Logos;

#[derive(Logos, Debug, PartialEq, Clone)]
#[logos(skip r"[ \t\n\f]+", skip r"#.*\n?", error = LexicalError)]
pub enum Token {
    #[token("{")]
    LBrace,
    #[token("}")]
    RBrace,
    #[token("|")]
    OR,
    #[token(",")]
    COMMA,
    #[token("::=")]
    Def,
    #[token(";")]
    Semi,
    #[regex("[0-9]+", |lex| lex.slice().parse::<u32>())]
    Int(u32),
    #[regex("<(?:[a-zA-Z]*)>", |lex| lex.slice()[1..lex.slice().len() - 1].to_string())]
    NonTerminal(String),
    #[regex(r#"'(?:\\'|[^(')])*'|"(?:\\"|[^(")])*""#, |lex| lex.slice()[1..lex.slice().len() - 1].to_string())]
    Terminal(String),
}

#[derive(Default, Debug, Clone, PartialEq)]
pub enum LexicalError {
    InvalidInteger(ParseIntError),
    #[default]
    InvalidToken,
}

impl From<ParseIntError> for LexicalError {
    fn from(err: ParseIntError) -> Self {
        LexicalError::InvalidInteger(err)
    }
}

impl Display for LexicalError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            LexicalError::InvalidInteger(err) => write!(f, "invalid integer: {}", err),
            LexicalError::InvalidToken => write!(f, "invalid token"),
        }
    }
}
