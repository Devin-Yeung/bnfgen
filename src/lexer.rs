use logos::{Logos, SpannedIter};

use crate::token::{LexicalError, Token};

pub type Spanned<Tok, Loc, Error> = Result<(Loc, Tok, Loc), Error>;

pub struct Lexer<'input> {
    // instead of an iterator over characters, we have a token iterator
    token_stream: SpannedIter<'input, Token>,
}

impl<'input> Lexer<'input> {
    pub fn new(input: &'input str) -> Self {
        Self {
            token_stream: Token::lexer(input).spanned(),
        }
    }
}

impl<'input> Iterator for Lexer<'input> {
    type Item = Spanned<Token, usize, LexicalError>;

    fn next(&mut self) -> Option<Self::Item> {
        self.token_stream
            .next()
            .map(|(token, span)| Ok((span.start, token?, span.end)))
    }
}

#[cfg(test)]
mod test {
    #[test]
    fn it_works() {
        let input = r#"
            <E> ::= <T> | <E> "+" <T> ';' "\"" '\'' ;
        "#;
        let mut lexer = super::Lexer::new(input);
        let tokens: Vec<_> = lexer.collect();
        insta::assert_debug_snapshot!(tokens);
    }
}
