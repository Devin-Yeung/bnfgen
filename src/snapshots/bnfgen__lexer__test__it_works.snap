---
source: src/lexer.rs
expression: tokens
---
[
    Ok(
        (
            1,
            NonTerminal(
                "syntax",
            ),
            9,
        ),
    ),
    Ok(
        (
            18,
            Def,
            21,
        ),
    ),
    Ok(
        (
            22,
            NonTerminal(
                "rule",
            ),
            28,
        ),
    ),
    Ok(
        (
            29,
            Or,
            30,
        ),
    ),
    Ok(
        (
            31,
            NonTerminal(
                "rule",
            ),
            37,
        ),
    ),
    Ok(
        (
            38,
            NonTerminal(
                "syntax",
            ),
            46,
        ),
    ),
    Ok(
        (
            47,
            Semi,
            48,
        ),
    ),
    Ok(
        (
            50,
            NonTerminal(
                "rule",
            ),
            56,
        ),
    ),
    Ok(
        (
            67,
            Def,
            70,
        ),
    ),
    Ok(
        (
            71,
            NonTerminal(
                "opt-whitespace",
            ),
            87,
        ),
    ),
    Ok(
        (
            88,
            Terminal(
                "<",
            ),
            91,
        ),
    ),
    Ok(
        (
            92,
            NonTerminal(
                "rule-name",
            ),
            103,
        ),
    ),
    Ok(
        (
            104,
            Terminal(
                ">",
            ),
            107,
        ),
    ),
    Ok(
        (
            108,
            NonTerminal(
                "opt-whitespace",
            ),
            124,
        ),
    ),
    Ok(
        (
            125,
            Terminal(
                "::=",
            ),
            130,
        ),
    ),
    Ok(
        (
            131,
            NonTerminal(
                "opt-whitespace",
            ),
            147,
        ),
    ),
    Ok(
        (
            148,
            NonTerminal(
                "expression",
            ),
            160,
        ),
    ),
    Ok(
        (
            161,
            NonTerminal(
                "line-end",
            ),
            171,
        ),
    ),
    Ok(
        (
            172,
            Semi,
            173,
        ),
    ),
    Ok(
        (
            175,
            NonTerminal(
                "opt-whitespace",
            ),
            191,
        ),
    ),
    Ok(
        (
            192,
            Def,
            195,
        ),
    ),
    Ok(
        (
            196,
            Terminal(
                " ",
            ),
            199,
        ),
    ),
    Ok(
        (
            200,
            NonTerminal(
                "opt-whitespace",
            ),
            216,
        ),
    ),
    Ok(
        (
            217,
            Or,
            218,
        ),
    ),
    Ok(
        (
            219,
            Terminal(
                "",
            ),
            221,
        ),
    ),
    Ok(
        (
            222,
            Semi,
            223,
        ),
    ),
    Ok(
        (
            225,
            NonTerminal(
                "expression",
            ),
            237,
        ),
    ),
    Ok(
        (
            242,
            Def,
            245,
        ),
    ),
    Ok(
        (
            246,
            NonTerminal(
                "list",
            ),
            252,
        ),
    ),
    Ok(
        (
            253,
            Or,
            254,
        ),
    ),
    Ok(
        (
            255,
            NonTerminal(
                "list",
            ),
            261,
        ),
    ),
    Ok(
        (
            262,
            NonTerminal(
                "opt-whitespace",
            ),
            278,
        ),
    ),
    Ok(
        (
            279,
            Terminal(
                "|",
            ),
            282,
        ),
    ),
    Ok(
        (
            283,
            NonTerminal(
                "opt-whitespace",
            ),
            299,
        ),
    ),
    Ok(
        (
            300,
            NonTerminal(
                "expression",
            ),
            312,
        ),
    ),
    Ok(
        (
            313,
            Semi,
            314,
        ),
    ),
    Ok(
        (
            316,
            NonTerminal(
                "line-end",
            ),
            326,
        ),
    ),
    Ok(
        (
            333,
            Def,
            336,
        ),
    ),
    Ok(
        (
            337,
            NonTerminal(
                "opt-whitespace",
            ),
            353,
        ),
    ),
    Ok(
        (
            354,
            NonTerminal(
                "EOL",
            ),
            359,
        ),
    ),
    Ok(
        (
            360,
            Or,
            361,
        ),
    ),
    Ok(
        (
            362,
            NonTerminal(
                "line-end",
            ),
            372,
        ),
    ),
    Ok(
        (
            373,
            NonTerminal(
                "line-end",
            ),
            383,
        ),
    ),
    Ok(
        (
            384,
            Semi,
            385,
        ),
    ),
    Ok(
        (
            387,
            NonTerminal(
                "list",
            ),
            393,
        ),
    ),
    Ok(
        (
            404,
            Def,
            407,
        ),
    ),
    Ok(
        (
            408,
            NonTerminal(
                "term",
            ),
            414,
        ),
    ),
    Ok(
        (
            415,
            Or,
            416,
        ),
    ),
    Ok(
        (
            417,
            NonTerminal(
                "term",
            ),
            423,
        ),
    ),
    Ok(
        (
            424,
            NonTerminal(
                "opt-whitespace",
            ),
            440,
        ),
    ),
    Ok(
        (
            441,
            NonTerminal(
                "list",
            ),
            447,
        ),
    ),
    Ok(
        (
            448,
            Semi,
            449,
        ),
    ),
    Ok(
        (
            451,
            NonTerminal(
                "term",
            ),
            457,
        ),
    ),
    Ok(
        (
            468,
            Def,
            471,
        ),
    ),
    Ok(
        (
            472,
            NonTerminal(
                "literal",
            ),
            481,
        ),
    ),
    Ok(
        (
            482,
            Or,
            483,
        ),
    ),
    Ok(
        (
            484,
            Terminal(
                "<",
            ),
            487,
        ),
    ),
    Ok(
        (
            488,
            NonTerminal(
                "rule-name",
            ),
            499,
        ),
    ),
    Ok(
        (
            500,
            Terminal(
                ">",
            ),
            503,
        ),
    ),
    Ok(
        (
            504,
            Semi,
            505,
        ),
    ),
    Ok(
        (
            507,
            NonTerminal(
                "literal",
            ),
            516,
        ),
    ),
    Ok(
        (
            524,
            Def,
            527,
        ),
    ),
    Ok(
        (
            528,
            Terminal(
                "\"",
            ),
            531,
        ),
    ),
    Ok(
        (
            532,
            NonTerminal(
                "text1",
            ),
            539,
        ),
    ),
    Ok(
        (
            540,
            Terminal(
                "\"",
            ),
            543,
        ),
    ),
    Ok(
        (
            544,
            Or,
            545,
        ),
    ),
    Ok(
        (
            546,
            Terminal(
                "'",
            ),
            549,
        ),
    ),
    Ok(
        (
            550,
            NonTerminal(
                "text2",
            ),
            557,
        ),
    ),
    Ok(
        (
            558,
            Terminal(
                "'",
            ),
            561,
        ),
    ),
    Ok(
        (
            562,
            Semi,
            563,
        ),
    ),
    Ok(
        (
            565,
            NonTerminal(
                "text1",
            ),
            572,
        ),
    ),
    Ok(
        (
            582,
            Def,
            585,
        ),
    ),
    Ok(
        (
            586,
            Terminal(
                "",
            ),
            588,
        ),
    ),
    Ok(
        (
            589,
            Or,
            590,
        ),
    ),
    Ok(
        (
            591,
            NonTerminal(
                "character1",
            ),
            603,
        ),
    ),
    Ok(
        (
            604,
            NonTerminal(
                "text1",
            ),
            611,
        ),
    ),
    Ok(
        (
            612,
            Semi,
            613,
        ),
    ),
    Ok(
        (
            615,
            NonTerminal(
                "text2",
            ),
            622,
        ),
    ),
    Ok(
        (
            632,
            Def,
            635,
        ),
    ),
    Ok(
        (
            636,
            Terminal(
                "",
            ),
            638,
        ),
    ),
    Ok(
        (
            639,
            Or,
            640,
        ),
    ),
    Ok(
        (
            641,
            NonTerminal(
                "character2",
            ),
            653,
        ),
    ),
    Ok(
        (
            654,
            NonTerminal(
                "text2",
            ),
            661,
        ),
    ),
    Ok(
        (
            662,
            Semi,
            663,
        ),
    ),
    Ok(
        (
            665,
            NonTerminal(
                "character",
            ),
            676,
        ),
    ),
    Ok(
        (
            682,
            Def,
            685,
        ),
    ),
    Ok(
        (
            686,
            NonTerminal(
                "letter",
            ),
            694,
        ),
    ),
    Ok(
        (
            695,
            Or,
            696,
        ),
    ),
    Ok(
        (
            697,
            NonTerminal(
                "digit",
            ),
            704,
        ),
    ),
    Ok(
        (
            705,
            Or,
            706,
        ),
    ),
    Ok(
        (
            707,
            NonTerminal(
                "symbol",
            ),
            715,
        ),
    ),
    Ok(
        (
            716,
            Semi,
            717,
        ),
    ),
    Ok(
        (
            719,
            NonTerminal(
                "letter",
            ),
            727,
        ),
    ),
    Ok(
        (
            736,
            Def,
            739,
        ),
    ),
    Ok(
        (
            740,
            Terminal(
                "A",
            ),
            743,
        ),
    ),
    Ok(
        (
            744,
            Or,
            745,
        ),
    ),
    Ok(
        (
            746,
            Terminal(
                "B",
            ),
            749,
        ),
    ),
    Ok(
        (
            750,
            Or,
            751,
        ),
    ),
    Ok(
        (
            752,
            Terminal(
                "C",
            ),
            755,
        ),
    ),
    Ok(
        (
            756,
            Or,
            757,
        ),
    ),
    Ok(
        (
            758,
            Terminal(
                "D",
            ),
            761,
        ),
    ),
    Ok(
        (
            762,
            Or,
            763,
        ),
    ),
    Ok(
        (
            764,
            Terminal(
                "E",
            ),
            767,
        ),
    ),
    Ok(
        (
            768,
            Or,
            769,
        ),
    ),
    Ok(
        (
            770,
            Terminal(
                "F",
            ),
            773,
        ),
    ),
    Ok(
        (
            774,
            Or,
            775,
        ),
    ),
    Ok(
        (
            776,
            Terminal(
                "G",
            ),
            779,
        ),
    ),
    Ok(
        (
            780,
            Or,
            781,
        ),
    ),
    Ok(
        (
            782,
            Terminal(
                "H",
            ),
            785,
        ),
    ),
    Ok(
        (
            786,
            Or,
            787,
        ),
    ),
    Ok(
        (
            788,
            Terminal(
                "I",
            ),
            791,
        ),
    ),
    Ok(
        (
            792,
            Or,
            793,
        ),
    ),
    Ok(
        (
            794,
            Terminal(
                "J",
            ),
            797,
        ),
    ),
    Ok(
        (
            798,
            Or,
            799,
        ),
    ),
    Ok(
        (
            800,
            Terminal(
                "K",
            ),
            803,
        ),
    ),
    Ok(
        (
            804,
            Or,
            805,
        ),
    ),
    Ok(
        (
            806,
            Terminal(
                "L",
            ),
            809,
        ),
    ),
    Ok(
        (
            810,
            Or,
            811,
        ),
    ),
    Ok(
        (
            812,
            Terminal(
                "M",
            ),
            815,
        ),
    ),
    Ok(
        (
            816,
            Or,
            817,
        ),
    ),
    Ok(
        (
            818,
            Terminal(
                "N",
            ),
            821,
        ),
    ),
    Ok(
        (
            822,
            Or,
            823,
        ),
    ),
    Ok(
        (
            824,
            Terminal(
                "O",
            ),
            827,
        ),
    ),
    Ok(
        (
            828,
            Or,
            829,
        ),
    ),
    Ok(
        (
            830,
            Terminal(
                "P",
            ),
            833,
        ),
    ),
    Ok(
        (
            834,
            Or,
            835,
        ),
    ),
    Ok(
        (
            836,
            Terminal(
                "Q",
            ),
            839,
        ),
    ),
    Ok(
        (
            840,
            Or,
            841,
        ),
    ),
    Ok(
        (
            842,
            Terminal(
                "R",
            ),
            845,
        ),
    ),
    Ok(
        (
            846,
            Or,
            847,
        ),
    ),
    Ok(
        (
            848,
            Terminal(
                "S",
            ),
            851,
        ),
    ),
    Ok(
        (
            852,
            Or,
            853,
        ),
    ),
    Ok(
        (
            854,
            Terminal(
                "T",
            ),
            857,
        ),
    ),
    Ok(
        (
            858,
            Or,
            859,
        ),
    ),
    Ok(
        (
            860,
            Terminal(
                "U",
            ),
            863,
        ),
    ),
    Ok(
        (
            864,
            Or,
            865,
        ),
    ),
    Ok(
        (
            866,
            Terminal(
                "V",
            ),
            869,
        ),
    ),
    Ok(
        (
            870,
            Or,
            871,
        ),
    ),
    Ok(
        (
            872,
            Terminal(
                "W",
            ),
            875,
        ),
    ),
    Ok(
        (
            876,
            Or,
            877,
        ),
    ),
    Ok(
        (
            878,
            Terminal(
                "X",
            ),
            881,
        ),
    ),
    Ok(
        (
            882,
            Or,
            883,
        ),
    ),
    Ok(
        (
            884,
            Terminal(
                "Y",
            ),
            887,
        ),
    ),
    Ok(
        (
            888,
            Or,
            889,
        ),
    ),
    Ok(
        (
            890,
            Terminal(
                "Z",
            ),
            893,
        ),
    ),
    Ok(
        (
            894,
            Or,
            895,
        ),
    ),
    Ok(
        (
            896,
            Terminal(
                "a",
            ),
            899,
        ),
    ),
    Ok(
        (
            900,
            Or,
            901,
        ),
    ),
    Ok(
        (
            902,
            Terminal(
                "b",
            ),
            905,
        ),
    ),
    Ok(
        (
            906,
            Or,
            907,
        ),
    ),
    Ok(
        (
            908,
            Terminal(
                "c",
            ),
            911,
        ),
    ),
    Ok(
        (
            912,
            Or,
            913,
        ),
    ),
    Ok(
        (
            914,
            Terminal(
                "d",
            ),
            917,
        ),
    ),
    Ok(
        (
            918,
            Or,
            919,
        ),
    ),
    Ok(
        (
            920,
            Terminal(
                "e",
            ),
            923,
        ),
    ),
    Ok(
        (
            924,
            Or,
            925,
        ),
    ),
    Ok(
        (
            926,
            Terminal(
                "f",
            ),
            929,
        ),
    ),
    Ok(
        (
            930,
            Or,
            931,
        ),
    ),
    Ok(
        (
            932,
            Terminal(
                "g",
            ),
            935,
        ),
    ),
    Ok(
        (
            936,
            Or,
            937,
        ),
    ),
    Ok(
        (
            938,
            Terminal(
                "h",
            ),
            941,
        ),
    ),
    Ok(
        (
            942,
            Or,
            943,
        ),
    ),
    Ok(
        (
            944,
            Terminal(
                "i",
            ),
            947,
        ),
    ),
    Ok(
        (
            948,
            Or,
            949,
        ),
    ),
    Ok(
        (
            950,
            Terminal(
                "j",
            ),
            953,
        ),
    ),
    Ok(
        (
            954,
            Or,
            955,
        ),
    ),
    Ok(
        (
            956,
            Terminal(
                "k",
            ),
            959,
        ),
    ),
    Ok(
        (
            960,
            Or,
            961,
        ),
    ),
    Ok(
        (
            962,
            Terminal(
                "l",
            ),
            965,
        ),
    ),
    Ok(
        (
            966,
            Or,
            967,
        ),
    ),
    Ok(
        (
            968,
            Terminal(
                "m",
            ),
            971,
        ),
    ),
    Ok(
        (
            972,
            Or,
            973,
        ),
    ),
    Ok(
        (
            974,
            Terminal(
                "n",
            ),
            977,
        ),
    ),
    Ok(
        (
            978,
            Or,
            979,
        ),
    ),
    Ok(
        (
            980,
            Terminal(
                "o",
            ),
            983,
        ),
    ),
    Ok(
        (
            984,
            Or,
            985,
        ),
    ),
    Ok(
        (
            986,
            Terminal(
                "p",
            ),
            989,
        ),
    ),
    Ok(
        (
            990,
            Or,
            991,
        ),
    ),
    Ok(
        (
            992,
            Terminal(
                "q",
            ),
            995,
        ),
    ),
    Ok(
        (
            996,
            Or,
            997,
        ),
    ),
    Ok(
        (
            998,
            Terminal(
                "r",
            ),
            1001,
        ),
    ),
    Ok(
        (
            1002,
            Or,
            1003,
        ),
    ),
    Ok(
        (
            1004,
            Terminal(
                "s",
            ),
            1007,
        ),
    ),
    Ok(
        (
            1008,
            Or,
            1009,
        ),
    ),
    Ok(
        (
            1010,
            Terminal(
                "t",
            ),
            1013,
        ),
    ),
    Ok(
        (
            1014,
            Or,
            1015,
        ),
    ),
    Ok(
        (
            1016,
            Terminal(
                "u",
            ),
            1019,
        ),
    ),
    Ok(
        (
            1020,
            Or,
            1021,
        ),
    ),
    Ok(
        (
            1022,
            Terminal(
                "v",
            ),
            1025,
        ),
    ),
    Ok(
        (
            1026,
            Or,
            1027,
        ),
    ),
    Ok(
        (
            1028,
            Terminal(
                "w",
            ),
            1031,
        ),
    ),
    Ok(
        (
            1032,
            Or,
            1033,
        ),
    ),
    Ok(
        (
            1034,
            Terminal(
                "x",
            ),
            1037,
        ),
    ),
    Ok(
        (
            1038,
            Or,
            1039,
        ),
    ),
    Ok(
        (
            1040,
            Terminal(
                "y",
            ),
            1043,
        ),
    ),
    Ok(
        (
            1044,
            Or,
            1045,
        ),
    ),
    Ok(
        (
            1046,
            Terminal(
                "z",
            ),
            1049,
        ),
    ),
    Ok(
        (
            1050,
            Semi,
            1051,
        ),
    ),
    Ok(
        (
            1053,
            NonTerminal(
                "digit",
            ),
            1060,
        ),
    ),
    Ok(
        (
            1070,
            Def,
            1073,
        ),
    ),
    Ok(
        (
            1074,
            Terminal(
                "0",
            ),
            1077,
        ),
    ),
    Ok(
        (
            1078,
            Or,
            1079,
        ),
    ),
    Ok(
        (
            1080,
            Terminal(
                "1",
            ),
            1083,
        ),
    ),
    Ok(
        (
            1084,
            Or,
            1085,
        ),
    ),
    Ok(
        (
            1086,
            Terminal(
                "2",
            ),
            1089,
        ),
    ),
    Ok(
        (
            1090,
            Or,
            1091,
        ),
    ),
    Ok(
        (
            1092,
            Terminal(
                "3",
            ),
            1095,
        ),
    ),
    Ok(
        (
            1096,
            Or,
            1097,
        ),
    ),
    Ok(
        (
            1098,
            Terminal(
                "4",
            ),
            1101,
        ),
    ),
    Ok(
        (
            1102,
            Or,
            1103,
        ),
    ),
    Ok(
        (
            1104,
            Terminal(
                "5",
            ),
            1107,
        ),
    ),
    Ok(
        (
            1108,
            Or,
            1109,
        ),
    ),
    Ok(
        (
            1110,
            Terminal(
                "6",
            ),
            1113,
        ),
    ),
    Ok(
        (
            1114,
            Or,
            1115,
        ),
    ),
    Ok(
        (
            1116,
            Terminal(
                "7",
            ),
            1119,
        ),
    ),
    Ok(
        (
            1120,
            Or,
            1121,
        ),
    ),
    Ok(
        (
            1122,
            Terminal(
                "8",
            ),
            1125,
        ),
    ),
    Ok(
        (
            1126,
            Or,
            1127,
        ),
    ),
    Ok(
        (
            1128,
            Terminal(
                "9",
            ),
            1131,
        ),
    ),
    Ok(
        (
            1132,
            Semi,
            1133,
        ),
    ),
    Ok(
        (
            1135,
            NonTerminal(
                "symbol",
            ),
            1143,
        ),
    ),
    Ok(
        (
            1152,
            Def,
            1155,
        ),
    ),
    Ok(
        (
            1156,
            Terminal(
                "|",
            ),
            1159,
        ),
    ),
    Ok(
        (
            1160,
            Or,
            1161,
        ),
    ),
    Ok(
        (
            1162,
            Terminal(
                " ",
            ),
            1165,
        ),
    ),
    Ok(
        (
            1166,
            Or,
            1167,
        ),
    ),
    Ok(
        (
            1168,
            Terminal(
                "!",
            ),
            1171,
        ),
    ),
    Ok(
        (
            1172,
            Or,
            1173,
        ),
    ),
    Ok(
        (
            1174,
            Terminal(
                "#",
            ),
            1177,
        ),
    ),
    Ok(
        (
            1178,
            Or,
            1179,
        ),
    ),
    Ok(
        (
            1180,
            Terminal(
                "$",
            ),
            1183,
        ),
    ),
    Ok(
        (
            1184,
            Or,
            1185,
        ),
    ),
    Ok(
        (
            1186,
            Terminal(
                "%",
            ),
            1189,
        ),
    ),
    Ok(
        (
            1190,
            Or,
            1191,
        ),
    ),
    Ok(
        (
            1192,
            Terminal(
                "&",
            ),
            1195,
        ),
    ),
    Ok(
        (
            1196,
            Or,
            1197,
        ),
    ),
    Ok(
        (
            1198,
            Terminal(
                "(",
            ),
            1201,
        ),
    ),
    Ok(
        (
            1202,
            Or,
            1203,
        ),
    ),
    Ok(
        (
            1204,
            Terminal(
                ")",
            ),
            1207,
        ),
    ),
    Ok(
        (
            1208,
            Or,
            1209,
        ),
    ),
    Ok(
        (
            1210,
            Terminal(
                "*",
            ),
            1213,
        ),
    ),
    Ok(
        (
            1214,
            Or,
            1215,
        ),
    ),
    Ok(
        (
            1216,
            Terminal(
                "+",
            ),
            1219,
        ),
    ),
    Ok(
        (
            1220,
            Or,
            1221,
        ),
    ),
    Ok(
        (
            1222,
            Terminal(
                ",",
            ),
            1225,
        ),
    ),
    Ok(
        (
            1226,
            Or,
            1227,
        ),
    ),
    Ok(
        (
            1228,
            Terminal(
                "-",
            ),
            1231,
        ),
    ),
    Ok(
        (
            1232,
            Or,
            1233,
        ),
    ),
    Ok(
        (
            1234,
            Terminal(
                ".",
            ),
            1237,
        ),
    ),
    Ok(
        (
            1238,
            Or,
            1239,
        ),
    ),
    Ok(
        (
            1240,
            Terminal(
                "/",
            ),
            1243,
        ),
    ),
    Ok(
        (
            1244,
            Or,
            1245,
        ),
    ),
    Ok(
        (
            1246,
            Terminal(
                ":",
            ),
            1249,
        ),
    ),
    Ok(
        (
            1250,
            Or,
            1251,
        ),
    ),
    Ok(
        (
            1252,
            Terminal(
                ";",
            ),
            1255,
        ),
    ),
    Ok(
        (
            1256,
            Or,
            1257,
        ),
    ),
    Ok(
        (
            1258,
            Terminal(
                ">",
            ),
            1261,
        ),
    ),
    Ok(
        (
            1262,
            Or,
            1263,
        ),
    ),
    Ok(
        (
            1264,
            Terminal(
                "=",
            ),
            1267,
        ),
    ),
    Ok(
        (
            1268,
            Or,
            1269,
        ),
    ),
    Ok(
        (
            1270,
            Terminal(
                "<",
            ),
            1273,
        ),
    ),
    Ok(
        (
            1274,
            Or,
            1275,
        ),
    ),
    Ok(
        (
            1276,
            Terminal(
                "?",
            ),
            1279,
        ),
    ),
    Ok(
        (
            1280,
            Or,
            1281,
        ),
    ),
    Ok(
        (
            1282,
            Terminal(
                "@",
            ),
            1285,
        ),
    ),
    Ok(
        (
            1286,
            Or,
            1287,
        ),
    ),
    Ok(
        (
            1288,
            Terminal(
                "[",
            ),
            1291,
        ),
    ),
    Ok(
        (
            1292,
            Or,
            1293,
        ),
    ),
    Ok(
        (
            1294,
            Terminal(
                "\\",
            ),
            1297,
        ),
    ),
    Ok(
        (
            1298,
            Or,
            1299,
        ),
    ),
    Ok(
        (
            1300,
            Terminal(
                "]",
            ),
            1303,
        ),
    ),
    Ok(
        (
            1304,
            Or,
            1305,
        ),
    ),
    Ok(
        (
            1306,
            Terminal(
                "^",
            ),
            1309,
        ),
    ),
    Ok(
        (
            1310,
            Or,
            1311,
        ),
    ),
    Ok(
        (
            1312,
            Terminal(
                "_",
            ),
            1315,
        ),
    ),
    Ok(
        (
            1316,
            Or,
            1317,
        ),
    ),
    Ok(
        (
            1318,
            Terminal(
                "`",
            ),
            1321,
        ),
    ),
    Ok(
        (
            1322,
            Or,
            1323,
        ),
    ),
    Ok(
        (
            1324,
            Terminal(
                "{",
            ),
            1327,
        ),
    ),
    Ok(
        (
            1328,
            Or,
            1329,
        ),
    ),
    Ok(
        (
            1330,
            Terminal(
                "}",
            ),
            1333,
        ),
    ),
    Ok(
        (
            1334,
            Or,
            1335,
        ),
    ),
    Ok(
        (
            1336,
            Terminal(
                "~",
            ),
            1339,
        ),
    ),
    Ok(
        (
            1340,
            Semi,
            1341,
        ),
    ),
    Ok(
        (
            1343,
            NonTerminal(
                "character1",
            ),
            1355,
        ),
    ),
    Ok(
        (
            1360,
            Def,
            1363,
        ),
    ),
    Ok(
        (
            1364,
            NonTerminal(
                "character",
            ),
            1375,
        ),
    ),
    Ok(
        (
            1376,
            Or,
            1377,
        ),
    ),
    Ok(
        (
            1378,
            Terminal(
                "'",
            ),
            1381,
        ),
    ),
    Ok(
        (
            1382,
            Semi,
            1383,
        ),
    ),
    Ok(
        (
            1385,
            NonTerminal(
                "character2",
            ),
            1397,
        ),
    ),
    Ok(
        (
            1402,
            Def,
            1405,
        ),
    ),
    Ok(
        (
            1406,
            NonTerminal(
                "character",
            ),
            1417,
        ),
    ),
    Ok(
        (
            1418,
            Or,
            1419,
        ),
    ),
    Ok(
        (
            1420,
            Terminal(
                "\"",
            ),
            1423,
        ),
    ),
    Ok(
        (
            1424,
            Semi,
            1425,
        ),
    ),
    Ok(
        (
            1427,
            NonTerminal(
                "rule-name",
            ),
            1438,
        ),
    ),
    Ok(
        (
            1444,
            Def,
            1447,
        ),
    ),
    Ok(
        (
            1448,
            NonTerminal(
                "letter",
            ),
            1456,
        ),
    ),
    Ok(
        (
            1457,
            Or,
            1458,
        ),
    ),
    Ok(
        (
            1459,
            NonTerminal(
                "rule-name",
            ),
            1470,
        ),
    ),
    Ok(
        (
            1471,
            NonTerminal(
                "rule-char",
            ),
            1482,
        ),
    ),
    Ok(
        (
            1483,
            Semi,
            1484,
        ),
    ),
    Ok(
        (
            1486,
            NonTerminal(
                "rule-char",
            ),
            1497,
        ),
    ),
    Ok(
        (
            1503,
            Def,
            1506,
        ),
    ),
    Ok(
        (
            1507,
            NonTerminal(
                "letter",
            ),
            1515,
        ),
    ),
    Ok(
        (
            1516,
            Or,
            1517,
        ),
    ),
    Ok(
        (
            1518,
            NonTerminal(
                "digit",
            ),
            1525,
        ),
    ),
    Ok(
        (
            1526,
            Or,
            1527,
        ),
    ),
    Ok(
        (
            1528,
            Terminal(
                "-",
            ),
            1531,
        ),
    ),
    Ok(
        (
            1532,
            Semi,
            1533,
        ),
    ),
    Ok(
        (
            1535,
            NonTerminal(
                "EOL",
            ),
            1540,
        ),
    ),
    Ok(
        (
            1552,
            Def,
            1555,
        ),
    ),
    Ok(
        (
            1556,
            Terminal(
                ";",
            ),
            1559,
        ),
    ),
    Ok(
        (
            1560,
            Semi,
            1561,
        ),
    ),
]
